### ISSUE TITLE
Secure mempool-based payments on top of Grin implementation or ProteanCharm for MimbleWimble

### ISSUE THREAD

**Issue by pstavraetos** [mimblewimble/grin#3153]
Hello to everybody.

Right now there are only two variants embedded in protocol to send coins from wallet A to wallet B: 
file transfer (complicated enough for non-tech-users) and http(s) communication (more complicated, no good).

I propose one other simple idea, which, i believe, will not require alot of changes in the current GRIN implementation.

We have Wallet A (WA) which transfers N amount of GRINs to Wallet B (WB).
We introduce one dedicated part of mempool for interactive "wallet messages" (WM_state)

1. WB generates one temporary private key (pkt_b) on EC, after generates corresponding public key (pubkt_b).

2. User of WB sends pubkt_b to user of WA. This can be done in different ways (qr-code/sms/e-mail/ etc)

3. WA generates one temporary private key (pkt_a) on EC, after generates corresponding public key (pubkt_a),
then WA builds a partial transaction file sending N amount of GRINs, gzip's it, encrypts 
it with pubkt_b using ECIES scheme and in such a way we get PART1.

4. WA generates digital token <PART1, CurrentBlockHeight> and publishes it to network with aim
to be propagated.

5. Every GRIN node checks for digital tokens in WM_state which are older than M blocks and in such case
they are removed from WM_state. If we want a lifespan of each token to be 10 minutes then M=10, or 
we can choose an other M. Lifespan-based approach was choosen to avoid WM_state overflow and to clean 
WM_state from digital tokens which were "translated" in transactions successfully.

6. Every GRIN node checks for new received digital tokens in WM_state and tries to decrypt them using 
his pkt_* and if success then we go to step 7.

7. WB receives one new digital token and decrypts it successfully. Now WB has all needed information to
generate a corresponding response, after that a response should be gzip'ed, encrypted with ECIES using 
pubkt_a and in such a way we get PART2.

8. WB generates digital token <PART2, CurrentBlockHeight> and publishes it to network with aim
to be propagated.

9. Finalization: WA receives new digital token, decyphers it successfully and finalizes a transaction.

WM_state avoids introduction of new p2p layer using current mempool, lifespan of tokens avoids significant 
overflows of WM_state as well as gives to interacting wallets "some time" to be offline and, still, the
process of transaction construction comes to his finale success. Generation of temporary (each time random)
private keys avoids any reconstruction of tx-graph.

Also, to avoid WM_state flood, can be introduced some rule or similar: every wallet can send no more then
K transactions/each block, where K should be relatively small (3-5 should be enough), otherwise such wallet
will be banned for some time (e.g. for time of 1440 blocks).

Awaiting for discussion/critics/etc.

**Comment by lehnberg** [comment]
Hello, and thanks for this proposal!

Just a quick clarification:

> Generation of temporary (each time random) private keys avoids any reconstruction of tx-graph.

While you avoid constructing a meaningful graph of the "addresses" used in this scheme (i.e. `pubkt_a -spends-to-> pubkt_b`), this has no impact on the ability to construct a tx graph of _output commitments_, what you see as inputs & outputs here: https://grinscan.net/block/456135.

**Comment by pstavraetos** [comment]
Yes, but my goal was to show that adding of such tech on top of existing GRIN do not reveals graph of payments.

**Comment by pstavraetos** [comment]
The only one potential problematic case to break some of privacy i see now is: 1) Interested Persons puts their sniffer nodes as many as possible so that to have them direct peers of every known exchange where everybody can buy GRIN 2) A very little amount of transactions per block (ideal - no txs except coinbase) 3) Sniffer nodes detect some activity in for of outgoing digital tokens so they know that an exchange tries to interact with someone and they (by luck or suspicion) have their sniffers directly connected to wallet which waits propagation of digital token and ready to send PART2 back. As soon as sniffers detect PART2 <answer> digital token they know with high probability (still, they does not know actually the meaning of tokens) that user of wallet received some GRINs. 4) Using same approach to detect with whom that user will interact later, an Interested Persons will able to know truth. 5) CONCLUSION: to avoid potential risk some kind of IP obfuscation needed (TOR?)

**Comment by pstavraetos** [comment]
> this has no impact on the ability to construct a tx graph of _output commitments_

Hello!

I see this project uses MW & some zk-related tech to hide TXOs, maybe such approach would help grin protocol acheive the same? +"Ethereum 9¾ uses Mimblewimble transaction and it prevents printing money out of thin air." +"With spent tags(nullifier), Ethereum 9¾ completely hides where the inputs come from."  https://github.com/ethereum934/eth-mimblewimble
